packages/chopsticks/src/blockchain/index.ts

```diff
@@ -83,11 +86,23 @@ export class Blockchain {
     if (offchainWorker) {
       this.offchainWorker = new OffchainWorker()
     }
+
+    this.#maxMemoryBlockCount = maxMemoryBlockCount
   }
 
-  #registerBlock(block: Block) {
-    this.#blocksByNumber[block.number] = block
+  async #registerBlock(block: Block) {
+    // if exceed max memory block count, delete the oldest block
+    if (this.#blocksByNumber.size === this.#maxMemoryBlockCount) {
+      const firstKey = this.#blocksByNumber.keys().next().value
+      this.#blocksByNumber.delete(firstKey)
+    }
+    this.#blocksByNumber.set(block.number, block)
     this.#blocksByHash[block.hash] = block
+    // save to db
+    if (this.db) {
```

Comment:
is it more reasonable to save on close server or save on register block? i think save here is easier and also may have better user experience

Reply:
Save on register is fine. Otherwise we may have long shutdown time or lost data if server is forced killed.